python/pyspark/serializers.py

```diff
@@ -401,6 +427,22 @@ def __repr__(self):
         return "ArrowStreamPandasUDFSerializer"
 
 
+class InterleavedArrowStreamPandasSerializer(ArrowStreamPandasUDFSerializer):
+
+    def __init__(self, timezone, safecheck, assign_cols_by_name):
+        super(InterleavedArrowStreamPandasSerializer, self).__init__(timezone, safecheck, assign_cols_by_name)
+
+    def load_stream(self, stream):
+        """
+        Deserialize ArrowRecordBatches to an Arrow table and return as a list of pandas.Series.
+        """
+        reader = InterleavedArrowReader(stream)
```

Comment:
I'm not sure how this would work as the ArrowStreamSerializer wouldn't know how we signal the number of dataframes nor that 0 dataframes means end of stream?

Reply:
`ArrowStreamSerializer` doesn't need to know how many streams in total are being sent. Calling `ArrowStreamSerializer.load_stream` will return an iterator that loads 1 single Arrow stream and then stops. It's basically what you are already doing in `InterleavedArrowReader._read_df`. You would just wrap in a while loop. Taking what you have already, for example:

```python
 def load_stream(self, stream):
    dataframes_in_group = None

    while dataframes_in_group is None or dataframes_in_group > 0:
      dataframes_in_group = read_int(self._stream)

      if dataframes_in_group == 2:
          batch1 = [batch for batch in ArrowStreamSerializer.load_stream(stream)]
          batch2 = [batch for batch in ArrowStreamSerializer.load_stream(stream)]
          yield ([self.arrow_to_pandas(c) for c in pa.Table.from_batches(batch1).itercolumns()],
                 [self.arrow_to_pandas(c) for c in pa.Table.from_batches(batch2).itercolumns()])

      elif dataframes_in_group != 0:
          raise ValueError(
                'Received Invalid number of dataframes in group {0}'.format(dataframes_in_group))

```

I think this is what you already have, just condensed a little.

Reply:
doh! yes you're quick right- this is much better.